Specs/GCMV8.lean

```diff
@@ -0,0 +1,262 @@
+    (i : Nat) (acc : List (BitVec e)) : List (BitVec e) :=
+    if i < n/e then
+      let lo := i * e
+      let hi := lo + e - 1
+      have h₀ : e = hi - lo + 1 := by simp only [hi, lo]; omega
+      let part : BitVec e := h₀ ▸ extractLsb hi lo x
+      let newacc := part :: acc
+      splitTR x e h (i + 1) newacc
+    else acc
+  splitTR x e h 0 []
+
+example : split 0xabcd1234#32 8 (by omega) = [0xab#8, 0xcd#8, 0x12#8, 0x34#8] :=
+  by native_decide
+
+def hi (x : BitVec 128) : BitVec 64 :=
+  extractLsb 127 64 x
+
+def lo (x : BitVec 128) : BitVec 64 :=
+  extractLsb 63 0 x
+
+------------------------------------------------------------------------------
+-- Functions related to galois field operations
+
+/-- Performs multiplication of polynomials over GF(2).
+  This algorithm is a shift and add multiplication.
+  For each bit of y from highest to lowest, if the bit is set,
+  it performs a GF add (xor) of x onto the shifted accumulator.
+-/
+def pmult (x: BitVec (m + 1)) (y : BitVec (n + 1)) : BitVec (m + n + 1) :=
+  let rec pmultTR (x: BitVec (m + 1)) (y : BitVec (n + 1)) (i : Nat)
+    (acc : BitVec (m + n + 1)) : BitVec (m + n + 1) :=
+    if i < n + 1 then
+      let acc := acc <<< 1
+      have h : n + (m + 1) = m + n + 1 := by omega
+      let tmp := if getMsb y i then (BitVec.zero n) ++ x else h ▸ BitVec.zero (m + n + 1)
+      let acc := h ▸ acc ^^^ tmp
+      pmultTR x y (i + 1) (h ▸ acc)
+    else acc
+  pmultTR x y 0 (BitVec.zero (m + n + 1))
+
+example: pmult 0b1101#4 0b10#2 = 0b11010#5 := by native_decide
+
+/-- Degree of x. -/
+protected def degree (x : BitVec n) : Nat :=
+  let rec degreeTR (x : BitVec n) (n : Nat) : Nat :=
+    if n = 0 then 0
+    else if getLsb x n then n else degreeTR x (n - 1)
+  degreeTR x (n - 1)
+example: GCMV8.degree 0b0101#4 = 2 := by native_decide
+
+/-- Subtract x from y if y's x-degree-th bit is 1. -/
+protected def reduce (x : BitVec n) (y : BitVec n) : BitVec n :=
+  if getLsb y (GCMV8.degree x) then y ^^^ x else y
+
+/-- Performs division of polynomials over GF(2). -/
+def pdiv (x: BitVec n) (y : BitVec m) (h : 0 < m): BitVec n :=
+  let rec pdivTR (x : BitVec n) (y : BitVec m) (i : Nat) (z : BitVec m)
+    (acc : BitVec n) : BitVec n :=
+    if i < n then
+      have h2 : (n - i - 1) - (n - i - 1) + 1 = 1 := by omega
+      let xi : BitVec 1 := h2 ▸ extractLsb (n - i - 1) (n - i - 1) x
+      have h3 : m = m - 1 - 0 + 1 := by omega
+      let zi : BitVec m := h3 ▸ extractLsb (m - 1) 0 ((GCMV8.reduce y z) ++ xi)
+      have h1 : 1 = GCMV8.degree y - GCMV8.degree y + 1 := by omega
+      let bit : BitVec 1 := h1 ▸ extractLsb (GCMV8.degree y) (GCMV8.degree y) zi
+      have h4 : 1 = (n - i - 1) - (n - i - 1) + 1 := by omega
+      let newacc : BitVec n := BitVec.partInstall (n - i - 1) (n - i - 1) (h4 ▸ bit) acc
+      pdivTR x y (i + 1) zi newacc
+    else acc
+  pdivTR x y 0 (BitVec.zero m) (BitVec.zero n)
+
+example : pdiv 0b1101#4 0b10#2 (by omega) = 0b110#4 := by native_decide
```

Comment:
`rfl` works for all `pmod`/`pdiv` examples.

Reply:
Done.